### PR TITLE
BeOp Bid Adapter: prefer canonical URL when present & prepend protocol

### PR DESCRIPTION
**Pull request by cyppan**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

Prefer the canonical url when present when setting the payload url for the BeOp adapter.

We've noticed at least sometimes the protocol is missing (not sure where the problem is) so we prepend it when needed.

### CODE REVIEW

**Comment by patmmccann** [comment]
deepAccess(bidderRequest, 'refererInfo.canonicalUrl') || config.getConfig('pageUrl') doesnt make sense to me; as pageUrl affects deepAccess(bidderRequest, 'refererInfo.canonicalUrl')

**Comment by sebrobert** [comment]
> deepAccess(bidderRequest, 'refererInfo.canonicalUrl') || config.getConfig('pageUrl') doesnt make sense to me; as pageUrl affects deepAccess(bidderRequest, 'refererInfo.canonicalUrl')

Is that true for older version of prebid.js ? If we want this fix deployed in 6.x and 7.x versions, was that one already available ?

**Comment by dgirardi** [comment]
For 6 I would use `bidderRequest.refererInfo.canonicalUrl`, then `window.top.location.href`, then `window.location.href`. But I think you need to wrap the second in a try-catch - in my testing it will throw an exception if it crosses origins.

`bidderRequest.referer` *almost* does that, but it returns something else (the referer) if it's running from  a cross-origin frame. Better to avoid its use for now, that logic will be reworked in 7.

the reason `canonicalUrl` sometimes misses the protocol is likely that the publisher did not provide it in either `setConfig({pageUrl})` or `<link rel="canonical" href="<pageUrl>" />`.

**Comment by cyppan** [comment]
Thanks @dgirardi, I've modified following your advice.

**Comment by cyppan** [comment]
Not sure why the tests fail, seems to come from another adapter test file 🤔

**Review by ChrisHuie — APPROVED** [review]

**Comment by ChrisHuie** [comment]
Sorry, ignore my review. I was changing the title and tabbed too far

File: modules/beopBidAdapter.js [hunk 1]
```diff
@@ -141,4 +143,47 @@ function beOpRequestSlotsMaker(bid) {
...
+    return null;
+  }
+}
+
+function getPageUrl(refererInfo, window) {
+  refererInfo = refererInfo || getRefererInfo();
+  let pageUrl = refererInfo.canonicalUrl || safeDeepAccess(window, 'top.location.href') || deepAccess(window, 'location.href');
+  // Ensure the protocol is present (looks like sometimes the extracted pageUrl misses it)
+  if (pageUrl != null) {
+    const defaultProtocol = deepAccess(window, 'top.location.protocol') || deepAccess(window, 'location.protocol');
```
Comment (dgirardi) [inline modules/beopBidAdapter.js:183]: 'top.location.protocol' can also throw.

**Pull request merged**